### PR TITLE
Caching: Refactor enterprise query caching middleware to a wire service

### PR DESCRIPTION
**Pull request by mmandrus**
**What is this feature?**

This work converts the enterprise query caching implementation from middleware to a service injected by wire. The OSS implementation of this service does nothing, while the enterprise service functionality remains the same.

This new implementation will live behind a feature flag in the near term. Once we have verified its stability, we will remove the flag. 

**Why do we need this feature?**

 Benefits of doing this include:

- Reduce risk of breaking changes between enterprise and OSS
- Reduce duplicated logic (e.g. assessing DS permissions in the middleware)
- Performance & safety improvements (stop non-query API calls from passing through middleware)

**Who is this feature for?**

Enterprise customers using query caching, and OSS developers. 

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana-enterprise/issues/4666

**Special notes for your reviewer**:

Corresponding enterprise PR is here https://github.com/grafana/grafana-enterprise/pull/4851

The majority of interactions with the cache happen in the new plugin middleware, in order to consolidate things as best as possible. There is a follow-up task to fully consolidate the resource caching usage.

### CODE REVIEW

File: pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+	if reqCtx == nil {
+		return m.next.CallResource(ctx, req, sender)
+	}
+
+	// time how long this request takes
+	start := time.Now()
+
+	// First look in the resource cache if enabled
+	resp := m.caching.HandleResourceRequest(ctx, req)
+	ch := reqCtx.Resp.Header().Get(caching.XCacheHeader)
```
Comment (mmandrus) [inline pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go:105]: Not sure how I feel about just relying on the header here. Debating making a bool for found/missed as a return value of `HandleResourceRequest` and `HandleQueryRequest`. Looking for other opinions.

File: pkg/services/caching/caching_service_mock.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
+// Code generated by mockery v2.14.0. DO NOT EDIT.
```
[the remark this answers is not in the document]
Comment (mmandrus) [reply pkg/services/caching/caching_service_mock.go:1]: Oh, good to know, thanks! I will swap out my use of mockery then.

File: pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+type CachingMiddleware struct {
+	next    plugins.Client
+	caching caching.CachingService
+	log     log.Logger
+}
+
+// QueryData receives a data request and attempts to access results already stored in the cache for that request.
+// If data is found, it will return it immediately. Otherwise, it will perform the queries as usual, then write the response to the cache.
```
[the remark this answers is not in the document]
Comment (mmandrus) [reply pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go:41]: That's correct - for example, if an error occurs during the cache read, we wouldn't want to (or may not be able to) provide an update function. I tried to say this in the service documentation, but I will add a bit more wording.

File: pkg/services/caching/caching_service_mock.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (suntala) [reply pkg/services/caching/caching_service_mock.go:1]: I want to emphasise that the issue is still in the draft phase though. The plan is to recommend this change to other teams as well, but there is no consensus yet (see discussion [here](https://raintank-corp.slack.com/archives/CMSUYD6H3/p1680710481197259) for example).

[the remark this answers is not in the document]
Comment (mmandrus) [reply pkg/services/caching/caching_service_mock.go:1]: That's alright, I think it is probably better to make a test mock that is simple and just does what I need anyway.

File: pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+	"github.com/grafana/grafana/pkg/services/contexthandler"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// NewCachingMiddleware creates a new plugins.ClientMiddleware that will
+// attempt to read and write query results to the cache
+func NewCachingMiddleware(cachingService caching.CachingService) plugins.ClientMiddleware {
+	log := log.New("caching_middleware")
+	if err := prometheus.Register(QueryRequestHistogram); err != nil {
+		log.Error("error registering prometheus collector", "error", err)
```
[the remark this answers is not in the document]
Comment (mmandrus) [reply pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go]: Done on [d475b7f](https://github.com/grafana/grafana/pull/65616/commits/d475b7fbe9a00904dd405497b0deae70d7c0605c)

File: pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go (hunk 1 above)
Comment (PoorlyDefinedBehaviour) [reply pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go:105]: I agree that using the response header is not obvious. Returning a `bool` seems fine.

File: pkg/services/caching/service.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+	// It can be set to nil by the method implementation (if there is an error, for example), so it should be checked before being called.
+	UpdateCacheFn CacheQueryResponseFn
+}
+
+func ProvideCachingService() *OSSCachingService {
+	return &OSSCachingService{}
+}
+
+//go:generate mockery --name CachingService --structname FakeOSSCachingService --inpackage --filename caching_service_mock.go
+type CachingService interface {
```
Comment (mmandrus) [inline pkg/services/caching/service.go:33]: Received a question on this in slack that I'll address here for others:

Question: Why does func (m *CachingMiddleware) QueryData update the cache by calling cr.UpdateCacheFn(ctx, resp) but func (m *CachingMiddleware) CallResource expects the caller to update the cache?

Answer: 
There are two reasons it’s done this way, one is a design choice, and the other is because it’s kind of forced on us due to the plugin api.

Design choice: when the query cache handler func is called, we do a lot of work up front to validate the queries, retrieve the cache config, generate a key, etc. I thought it would be inefficient to do all that stuff twice, once for the cache retrieval, and another for the update. By returning a function, I can just do it all once and also give us better control over calls to update the cache in general.

Plugin API: there’s also a major difference in how the query and resource calls work. With the query call, you get back a single query response synchronously, so getting just that result and caching it is straightforward. With the resource API, there can be multiple responses, and they are sent from the plugin via a stream, which we read higher up the call hierarchy. If I wanted to return a cache func for the resource call, I would either have to read from the stream in the middleware, or pass that update func way back up the call stack - neither of these are very easily done using the plugin middleware

File: pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+// If the cache service is implemented, we capture the request duration as a metric. The service is expected to write any response headers.
+func (m *CachingMiddleware) QueryData(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
+	if req == nil {
+		return m.next.QueryData(ctx, req)
+	}
+
+	reqCtx := contexthandler.FromContext(ctx)
+	if reqCtx == nil {
+		return m.next.QueryData(ctx, req)
+	}
```
Comment (papagian) [inline pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go:51]: nit; I wonder if it makes sense to forward the request to the data source in these cases?

File: pkg/services/pluginsintegration/clientmiddleware/caching_middleware_test.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+			})
+
+			call.Run(func(args mock.Arguments) {
+				reqCtx.Resp.Header().Set(caching.XCacheHeader, caching.StatusMiss)
+			})
+
+			resp, err := cdt.Decorator.QueryData(req.Context(), qdr)
+			assert.NoError(t, err)
+			// Cache service is called again
+			cs.AssertNumberOfCalls(t, "HandleQueryRequest", 2)
```
Comment (papagian) [inline pkg/services/pluginsintegration/clientmiddleware/caching_middleware_test.go:91]: this check depends on the above test, therefore it fails if the test runs alone:
<details>
<summary>test output</summary>

```
~/go/src/github.com/grafana/grafana: go test -v ./pkg/services/pluginsintegration/clientmiddleware/... -run TestCachingMiddleware/When_QueryData_is_called/If_cache_returns_a_miss,_queries_are_issued_and_the_update_cache_function_is_called
=== RUN   TestCachingMiddleware
=== RUN   TestCachingMiddleware/When_QueryData_is_called
=== RUN   TestCachingMiddleware/When_QueryData_is_called/If_cache_returns_a_miss,_queries_are_issued_and_the_update_cache_function_is_called
    caching_middleware_test.go:91: 
                Error Trace:    /Users/josefk/go/src/github.com/grafana/grafana/pkg/services/pluginsintegration/clientmiddleware/caching_middleware_test.go:91
                Error:          Not equal: 
                                expected: 2
                                actual  : 1
                Test:           TestCachingMiddleware/When_QueryData_is_called/If_cache_returns_a_miss,_queries_are_issued_and_the_update_cache_function_is_called
                Messages:       Expected number of calls (2) does not match the actual number of calls (1).
=== NAME  TestCachingMiddleware/When_QueryData_is_called
    caching_service_mock.go:63: PASS:   HandleQueryRequest(*context.valueCtx,*backend.QueryDataRequest)
--- FAIL: TestCachingMiddleware (0.00s)
    --- FAIL: TestCachingMiddleware/When_QueryData_is_called (0.00s)
        --- FAIL: TestCachingMiddleware/When_QueryData_is_called/If_cache_returns_a_miss,_queries_are_issued_and_the_update_cache_function_is_called (0.00s)
FAIL
FAIL    github.com/grafana/grafana/pkg/services/pluginsintegration/clientmiddleware     0.369s
FAIL
```
</details>

it would be better if the tests were independent (create a new fake service for each test maybe?)

**Review by papagian — COMMENTED** [review]
backend-wise LGTM; I left a small comment for the tests

File: pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go (hunk 6 above)
Comment (marefr) [reply pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go:51]: In the case of alerting etc there will be no request context available

File: pkg/services/caching/service.go (hunk 5 above)
Comment (marefr) [reply pkg/services/caching/service.go:32]: Thanks. Had similar questions. It's unfortunate the call resource streaming logic lives in the API layer. It's quite a code smell. I think the following logic could be moved downwards to plugin client:
https://github.com/grafana/grafana/blob/1c3ce0735f3b6e5609c3a7d8988d7f0700ee75b1/pkg/api/plugin_resource.go#L167-L192

Plugin client already does some cleaning of HTTP response headers:
https://github.com/grafana/grafana/blob/1c3ce0735f3b6e5609c3a7d8988d7f0700ee75b1/pkg/plugins/manager/client/client.go#L96-L103

In call resource handle of caching client middleware you can implement similar `cachedSender` as above and do similar things with caching you're currently doing in plugin_resource.go

With that said, I think it could be a nice follow up task for the future.

Comment (marefr) [reply pkg/services/caching/service.go:32]: And in addition, moving things downstream is probably a task for @grafana/plugins-platform-backend to pick up eventually

File: pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+	// time how long this request takes
+	start := time.Now()
+
+	// First look in the query cache if enabled
+	hit, cr := m.caching.HandleQueryRequest(ctx, req)
+
+	defer func() {
+		// record request duration if caching was used
+		if ch := reqCtx.Resp.Header().Get(caching.XCacheHeader); ch != "" {
```
Comment (marefr) [inline pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go:61]: Looks a bit weird looking at HTTP response header. Doesn't this depend on the value of `hit`? Is it because that response header only is set if enterprise and query caching enabled?

File: pkg/services/pluginsintegration/clientmiddleware/caching_metrics.go [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+	Buckets:   []float64{.005, .01, .025, .05, .1, .25, .5, 1, 2.5, 5, 10, 25, 50, 100},
+}, []string{"datasource_type", "cache", "query_type"})
+
+var ResourceRequestHistogram = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+	Namespace: metrics.ExporterName,
+	Subsystem: "caching",
+	Name:      "resource_request_duration_seconds",
+	Help:      "histogram of grafana resource endpoint requests in seconds",
+	Buckets:   []float64{.005, .01, .025, .05, .1, .25, .5, 1, 2.5, 5, 10, 25, 50, 100},
+}, []string{"datasource_type", "cache"})
```
Comment (marefr) [inline pkg/services/pluginsintegration/clientmiddleware/caching_metrics.go:28]: Would suggest naming these metrics to make it clear they're query caching metrics.

File: pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go (hunk 1 above)
Comment (mmandrus) [reply pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go]: Done on [1bb82aa](https://github.com/grafana/grafana/pull/65616/commits/1bb82aa25e5da0d997a11adbf0341e6454bbea68)

File: pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+	start := time.Now()
+
+	// First look in the resource cache if enabled
+	hit, resp := m.caching.HandleResourceRequest(ctx, req)
+
+	defer func() {
+		// record request duration if caching was used
+		if ch := reqCtx.Resp.Header().Get(caching.XCacheHeader); ch != "" {
+			ResourceRequestHistogram.With(prometheus.Labels{
+				"datasource_type": req.PluginContext.DataSourceInstanceSettings.Type,
```
Comment (marefr) [inline pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go:109]: `req.PluginContext.DataSourceInstanceSettings` would be nil if request is made for an app plugin.

```suggestion
				"plugin_id": req.PluginContext.PluginID,
```

File: pkg/services/caching/caching_service_mock.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mmandrus) [reply pkg/services/caching/caching_service_mock.go]: Done on [1bb82aa](https://github.com/grafana/grafana/pull/65616/commits/1bb82aa25e5da0d997a11adbf0341e6454bbea68)

File: pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+		cr.UpdateCacheFn(ctx, resp)
+	}
+
+	return resp, err
+}
+
+// CallResource receives a resource request and attempts to access results already stored in the cache for that request.
+// If data is found, it will return it immediately. Otherwise, it will perform the request as usual. The caller of CallResource is expected to explicitly update the cache with any responses.
+// If the cache service is implemented, we capture the request duration as a metric. The service is expected to write any response headers.
+func (m *CachingMiddleware) CallResource(ctx context.Context, req *backend.CallResourceRequest, sender backend.CallResourceResponseSender) error {
```
Comment (marefr) [inline pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go:89]: This gets a bit weird/hard to follow now when (as I understand) the actual caching is made in api/plugin_resource.go, but the check to see if it's already cached is made here. That's not optimal.

File: pkg/services/pluginsintegration/clientmiddleware/caching_middleware_test.go (hunk 7 above)
Comment (mmandrus) [reply pkg/services/pluginsintegration/clientmiddleware/caching_middleware_test.go]: Good catch. I just made an update to remove mockery and Reset my fake service between tests, so this is fixed now. Just verified they run independently. [1bb82aa](https://github.com/grafana/grafana/pull/65616/commits/1bb82aa25e5da0d997a11adbf0341e6454bbea68)

File: pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go (hunk 6 above)
Comment (mmandrus) [reply pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go:51]: Thanks, wasn't sure if this could happen but I wanted to guard against it anyway. Good to know about alerting. I'll write a unit test around empty req contexts.

File: pkg/middleware/middleware.go [hunk 12]
```diff
@@ -21,8 +21,9 @@ var (
 	ReqOrgAdmin            = RoleAuth(org.RoleAdmin)
 )
 
-func HandleNoCacheHeader(ctx *contextmodel.ReqContext) {
-	ctx.SkipCache = ctx.Req.Header.Get("X-Grafana-NoCache") == "true"
+func HandleNoCacheHeaders(ctx *contextmodel.ReqContext) {
+	ctx.SkipDSCache = ctx.Req.Header.Get("X-Grafana-NoCache") == "true"
+	ctx.SkipQueryCache = ctx.Req.Header.Get("X-Cache-Skip") == "true"
```
Comment (marefr) [inline pkg/middleware/middleware.go:26]: Where's this used, enterprise?

**Review by marefr — COMMENTED** [review]
Looks quite good I think, but see comments.

File: pkg/services/pluginsintegration/clientmiddleware/caching_metrics.go (hunk 9 above)
Comment (marefr) [reply pkg/services/pluginsintegration/clientmiddleware/caching_metrics.go:28]: Sorry missed the subsystem :smile:

File: pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go (hunk 6 above)
Comment (mmandrus) [reply pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go:51]: Done on [3153299](https://github.com/grafana/grafana/pull/65616/commits/315329998a2713fbc39bd4201acfa1a307ca49ec)

File: pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go (hunk 8 above)
Comment (mmandrus) [reply pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go:61]: @PoorlyDefinedBehaviour shared a similar opinion above, so I changed the interface to return a bool instead [1bb82aa](https://github.com/grafana/grafana/pull/65616/commits/1bb82aa25e5da0d997a11adbf0341e6454bbea68). OSS no longer cares what headers are set.

File: pkg/middleware/middleware.go (hunk 12 above)
Comment (mmandrus) [reply pkg/middleware/middleware.go:26]: Yes - it seemed closely related enough that I decided to put it here, though I could just read this header from the requests each time in the enterprise impl. 

Do you think it would be sufficient to just make a comment here, or do you think I should remove this from OSS?

File: pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go (hunk 11 above)
Comment (mmandrus) [reply pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go:89]: Agreed, but not sure how to do it better right now, given the sender reading is done upstream.

Would it make sense to leave a comment here pointing to the caching call in `plugin_resource.go`, and create a follow-on task as you suggested above?

Comment (marefr) [reply pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go:89]: Yeah possibly

File: pkg/middleware/middleware.go (hunk 12 above)
Comment (marefr) [reply pkg/middleware/middleware.go:26]: `X-Grafana-NoCache` and `X-Cache-Skip` looks very similar, `SkipDSCache` and `SkipQueryCache` make things more understandable. Think it's fine to keep, but come comments could be good I think.

Comment (mmandrus) [reply pkg/middleware/middleware.go:28]: Added comments: [b51bfb5](https://github.com/grafana/grafana/pull/65616/commits/b51bfb54214a4c53c8f1cddaa84d35b9f83b7acd)

File: pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go (hunk 11 above)
Comment (mmandrus) [reply pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go:89]: I've left a comment [b7290fb](https://github.com/grafana/grafana/pull/65616/commits/b7290fbd6b7914bc2412b7f92465c5a28ba2220c)

And created a follow on issue https://github.com/grafana/grafana/issues/66297. Will prioritize this post-merge.

**Review by marefr — APPROVED** [review]
LGTM

**Review by leandro-deveikis — APPROVED** [review]
LGTM! :tada:
